types/supertest/index.d.ts

```diff
@@ -5,40 +5,36 @@
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // TypeScript Version: 3.0
 
-import * as superagent from "superagent"
+import * as superagent from 'superagent';
 
 export = supertest;
 
 declare function supertest(app: any): supertest.SuperTest<supertest.Test>;
 declare namespace supertest {
-    interface Response extends superagent.Response {
-    }
+    interface Response extends superagent.Response {}
 
-    interface Request extends superagent.SuperAgentRequest {
-    }
+    interface Request extends superagent.SuperAgentRequest {}
 
     type CallbackHandler = (err: any, res: Response) => void;
     interface Test extends superagent.SuperAgentRequest {
-      app?: any;
-      url: string;
-      serverAddress(app: any, path: string): string;
-      expect(status: number, callback?: CallbackHandler): this;
-      expect(status: number, body: any, callback?: CallbackHandler): this;
-      expect(checker: (res: Response) => any, callback?: CallbackHandler): this;
-      expect(body: string, callback?: CallbackHandler): this;
-      expect(body: RegExp, callback?: CallbackHandler): this;
-      expect(body: Object, callback?: CallbackHandler): this;
-      expect(field: string, val: string, callback?: CallbackHandler): this;
-      expect(field: string, val: RegExp, callback?: CallbackHandler): this;
-      end(callback?: CallbackHandler): this;
+        app?: any;
+        url: string;
+        serverAddress(app: any, path: string): string;
+        expect(status: number, callback?: CallbackHandler): this;
+        expect(status: number, body: any, callback?: CallbackHandler): this;
+        expect(checker: (res: Response) => any, callback?: CallbackHandler): this;
+        expect(body: string, callback?: CallbackHandler): this;
+        expect(body: RegExp, callback?: CallbackHandler): this;
+        expect(body: Object, callback?: CallbackHandler): this;
+        expect(field: string, val: string, callback?: CallbackHandler): this;
+        expect(field: string, val: RegExp, callback?: CallbackHandler): this;
+        end(callback?: CallbackHandler): this;
     }
 
     interface AgentOptions {
-      ca?: any;
-    }
-    function agent(app?: any, options?: AgentOptions): SuperTest<Test>;
-
-    interface SuperTest<T extends superagent.SuperAgentRequest> extends superagent.SuperAgent<T> {
+        ca?: any;
     }
+    function agent(app?: any, options?: AgentOptions): SuperTest<Test> & Request;
```

Comment:
From the docs you linked, it’s not clear to me that this is correct. The docs state

> The complete list of methods that the agent can use to set defaults is: use, on, once, set, query, type, accept, auth, withCredentials, sortQuery, retry, ok, redirects, timeout, buffer, serialize, parse, ca, key, pfx, cert.

But the `Request` type you‘re referencing is here: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/2811ee6409b192cd5c8e8d0827b227b43bf0d66b/types/superagent/index.d.ts#L131-L173

There is a fair amount of overlap, but there are methods `Request` has that aren’t listed in the docs (e.g. `abort()`), and methods listed in the docs that aren’t on `Request` (e.g. `sortQuery`).

Reply:
@andrewbranch - thanks for taking the time to thoughtfully review this change.

You're right, and I think either the superagent docs are out of sync with the code, or the superagent typings are incorrect.

Here's the definition for superagent.agent()
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/superagent/index.d.ts#L42
Note that it returns `Request`.  And so I though like supertests's `agent()` should do the same.

Reply:
Hmm, that does seem a little odd. I don’t have time to do a deeper investigation into this myself, and since this is a Popular Package (i.e. lots of downloads of the types on npm), we make an extra effort not to merge stuff that we’re not sure is right. If you have time to investigate and determine what the true types are, based on the runtime implementation, then we can move forward.

I also think a more conservative approach would be acceptable, where you make a new type with only the properties that the docs say should exist. It’s generally ok to have incomplete types, just not _incorrect_ types.

@typescript-bot might continue to close this PR until tomorrow—I’m working on a fix for that.

Reply:
Thanks.  I've added a custom type that explicitly lists the supported properties.  I found two that aren't listed in superagent's Request though: `once` and `sortQuery`.

Does this look OK?

Reply:
Apparently I can’t reopen this PR since you’ve pushed new changes to the branch after it was closed—can you open a new one? Sorry for the inconvenience!

Reply:
Sure thing! https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44420